qir/qat/Validator/Tests/Unit/RequiredQubitsAndResults.cpp

```diff
@@ -0,0 +1,153 @@
+    void errorNoQubitsPresent(llvm::Value*) override
+    {
+        raise_no_qubits_error_ = true;
+    }
+
+    void errorNoResultsPresent(llvm::Value*) override
+    {
+        raise_no_results_error_ = true;
+    }
+
+    bool raiseNoQubitsError() const
+    {
+        return raise_no_qubits_error_;
+    }
+
+    bool raiseNoResultsError() const
+    {
+        return raise_no_results_error_;
+    }
+
+  private:
+    bool raise_no_qubits_error_{false};
+    bool raise_no_results_error_{false};
+};
+
+void testSkeleton(String const& script, std::shared_ptr<MockLogger> const& logger)
+{
+    auto ir_manip = newIrManip(script);
+
+    auto generator = std::make_shared<ProfileGenerator>();
+
+    ConfigurationManager& configuration_manager = generator->configurationManager();
+    configuration_manager.addConfig<FactoryConfiguration>();
+    configuration_manager.addConfig<ValidationPassConfiguration>(
+        "validation-configuration", ValidationPassConfiguration::fromProfileName("generic"));
+
+    generator->setLogger(logger);
+    generator->setupDefaultComponentPipeline();
+
+    ParameterParser parser;
+    configuration_manager.setupArguments(parser);
+    char* argv[] = {"exe", "--requires-results", "--requires-qubits"};
+    parser.parseArgs(3, argv);
+
+    configuration_manager.configure(parser);
+    for (auto& prop : parser.unusedSettings())
+    {
+        llvm::errs() << "Unknown option or flag '" << prop
+                     << "'. If this is an experimental feature remember to add `--experimental` to enable it."
+                        "\n";
+    }
+
+    ir_manip->validateProfile(generator, "generic", false);
+}
+
+} // namespace
+
+// Single allocation with action and then release
+TEST(FunctionValidationPass, FailMissingResults)
+{
+    auto logger = std::make_shared<MockLogger>();
+    testSkeleton(
+        R"script(
+  tail call void @__quantum__qis__h__body(%Qubit* nonnull inttoptr (i64 2 to %Qubit*))
+  )script",
+        logger);
+    EXPECT_FALSE(logger->raiseNoQubitsError());
+    EXPECT_TRUE(logger->raiseNoResultsError());
+}
+
+TEST(FunctionValidationPass, FailMissingQubitsAndResults)
+{
+    auto logger = std::make_shared<MockLogger>();
+    testSkeleton(
+        R"script(  
+  )script",
+        logger);
+    EXPECT_TRUE(logger->raiseNoQubitsError());
+    EXPECT_TRUE(logger->raiseNoResultsError());
+}
+
+TEST(FunctionValidationPass, FailMissingQubits)
+{
+    auto logger = std::make_shared<MockLogger>();
+    testSkeleton(
+        R"script(  
+  tail call void @__quantum__reset__result(%Result* nonnull inttoptr (i64 2 to %Result*))    
+  )script",
+        logger);
+    EXPECT_TRUE(logger->raiseNoQubitsError());
+    EXPECT_FALSE(logger->raiseNoResultsError());
+}
+
+TEST(FunctionValidationPass, NothingMissing)
+{
+    auto logger = std::make_shared<MockLogger>();
+    testSkeleton(
+        R"script(  
+  tail call void @__quantum__reset__result(%Result* nonnull inttoptr (i64 2 to %Result*))    
```

Comment:
What is `__quantum__reset__result`? I can't say I've seen that before and it isn't `qis` or `rt` prefixed. Is it just an extern function to force result allocation in the test, and then you don't run the extern function check?

Reply:
It is just a made up function for the purpose of testing since I did not find a function that only takes a `Result*` as argument in the spec.

Reply:
I've changed the name to avoid any confusion.